### PR TITLE
Primary caching 17: timeless range

### LINKED ISSUE #4821: Primary cache: timeless support for ranges

### PR DESCRIPTION
**Pull request by teh-cmc**
Simply add a timeless path for the range cache, and actually only iterate over the range the user asked for (we were still blindly iterating over everything until now).

Also some very minimal clean up related to #4832, but we have a long way to go...
- #4832

---

- Fixes #4821 

---

Part of the primary caching series of PR (index search, joins, deserialization):
- #4592
- #4593
- #4659
- #4680 
- #4681
- #4698
- #4711
- #4712
- #4721 
- #4726 
- #4773
- #4784
- #4785
- #4793
- #4800
- #4851
- #4852
- #4853
- #4856

---

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4800/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4800/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4800/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4800)
- [Docs preview](https://rerun.io/preview/f9f71a7fabd3aa0c4d76d7425ac7dde76535bb31/docs) 
- [Examples preview](https://rerun.io/preview/f9f71a7fabd3aa0c4d76d7425ac7dde76535bb31/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_query_cache/src/range.rs [hunk 1]
```diff
@@ -171,16 +180,48 @@ macro_rules! impl_query_archetype_range {
             let mut range_callback = |query: &RangeQuery, range_cache: &mut crate::RangeCache| {
                 re_tracing::profile_scope!("range", format!("{query:?}"));
 
-                for reduced_query in range_cache.compute_queries(query) {
+                eprintln!("query 1: {query:?}");
```
Comment (teh-cmc) [inline crates/re_query_cache/src/range.rs:183]: nope

File: crates/re_query_cache/src/range.rs [hunk 2]
```diff
@@ -171,16 +180,48 @@ macro_rules! impl_query_archetype_range {
...
+
+                // NOTE: Same logic as what the store does.
+                if query.range.min <= TimeInt::MIN {
+                    let mut reduced_query = query.clone();
+                    reduced_query.range.max = TimeInt::MIN; // inclusive
+
+                    eprintln!("query timeless: {reduced_query:?}");
```
Comment (teh-cmc) [inline crates/re_query_cache/src/range.rs:190]: nope

File: crates/re_query_cache/src/range.rs [hunk 3]
```diff
@@ -104,19 +109,23 @@ macro_rules! impl_query_archetype_range {
                 ),
             ),
         {
-            let mut iter_results = |bucket: &crate::CacheBucket| -> crate::Result<()> {
+            let mut range_results =
+                |timeless: bool, bucket: &crate::CacheBucket, time_range: TimeRange| -> crate::Result<()> {
                 re_tracing::profile_scope!("iter");
 
+                let entry_range = bucket.entry_range(time_range);
+        dbg!(&entry_range);
```
Comment (teh-cmc) [inline crates/re_query_cache/src/range.rs:117]: nope

File: crates/re_query_cache/src/range.rs [hunk 4]
```diff
@@ -171,16 +180,48 @@ macro_rules! impl_query_archetype_range {
...
+                    if !range_cache.timeless.is_empty() {
+                        range_results(true, &range_cache.timeless, reduced_query.range)?;
+                    }
+                }
+
+
+                let mut query = query.clone();
+                query.range.min = TimeInt::max((TimeInt::MIN.as_i64() + 1).into(), query.range.min);
+
+                eprintln!("query 2: {query:?}");
```
Comment (teh-cmc) [inline crates/re_query_cache/src/range.rs:208]: nope

File: crates/re_query_cache/src/range.rs [hunk 5]
```diff
@@ -171,16 +180,48 @@ macro_rules! impl_query_archetype_range {
...
+
+                for reduced_query in range_cache.compute_queries(&query) {
+                    eprintln!("query 2: {reduced_query:?}");
```
Comment (teh-cmc) [inline crates/re_query_cache/src/range.rs:211]: nope

File: crates/re_query_cache/tests/range.rs [hunk 6]
```diff
@@ -212,6 +210,8 @@ fn timeless_range() {
 
     // --- Third test: `[-inf, +inf]` ---
 
+    eprintln!("XXXXXXXXXXXXXXXXXXXXXXXX");
+
```
Comment (teh-cmc) [inline crates/re_query_cache/tests/range.rs:214]: nope

File: crates/re_query_cache/tests/range.rs [hunk 7]
```diff
@@ -363,7 +363,7 @@ fn query_and_compare(store: &DataStore, query: &RangeQuery, ent_path: &EntityPat
         }
 
         // Keep this around for the next unlucky chap.
-        // eprintln!("(expected={expected_data_times:?}, uncached={uncached_data_times:?}, cached={cached_data_times:?})");
+        eprintln!("(expected={expected_data_times:?}, uncached={uncached_data_times:?}, cached={cached_data_times:?})");
```
Comment (teh-cmc) [inline crates/re_query_cache/tests/range.rs:366]: nope

**Review by jleibs — APPROVED** [review]

File: crates/re_query_cache/src/range.rs [hunk 8]
```diff
@@ -171,16 +179,41 @@ macro_rules! impl_query_archetype_range {
             let mut range_callback = |query: &RangeQuery, range_cache: &mut crate::RangeCache| {
                 re_tracing::profile_scope!("range", format!("{query:?}"));
 
-                for reduced_query in range_cache.compute_queries(query) {
+                // NOTE: Same logic as what the store does.
+                if query.range.min <= TimeInt::MIN {
+                    let mut reduced_query = query.clone();
+                    reduced_query.range.max = TimeInt::MIN; // inclusive
```
Comment (jleibs) [inline crates/re_query_cache/src/range.rs:185]: This line looks suspicious to me. Doesn't setting the max to the min make this an empty query?

Comment (teh-cmc) [reply crates/re_query_cache/src/range.rs:185]: It's inclusive so it yields `MIN..=MIN` = `[MIN]`

Comment (jleibs) [reply crates/re_query_cache/src/range.rs:185]: Got it -- some kind of comment here would be helpful explaining this is the reduced query corresponding to the timeless part of the data?

**Review by jleibs — APPROVED** [review]
:man_shrugging: still extremely hard to follow how this whole thing works, but if I assume everything outside of this is PR is correct I guess this makes sense.

File: crates/re_query_cache/src/range.rs (hunk 8 above)
Comment (teh-cmc) [reply crates/re_query_cache/src/range.rs:185]: will do
